Assignment6.java

```diff
@@ -77,24 +77,28 @@
      * @param indexMap the map from cars to slots
      * @return next car to be moved (0 in case of rearranged array)
      */
-    private static int nextCarToMove(int[] given, int[] desired, Map<Integer, Integer> indexMap) {
+    private static int nextCarToMove(int[] given, int[] desired, Map<Integer, Integer> indexMap, HashSet<Integer> misplacedCars) {
         //arrays of length 0
         if (given.length == 0) {
             return 0;
         }
         
         //car belonging to the current empty slot
         if (desired[indexMap.get(0)] != 0) {
+            // car will be moved in place next, remove it from set of misplaced cars
+            misplacedCars.remove(desired[indexMap.get(0)]);
             return desired[indexMap.get(0)];
+
         }
-        
+
         //empty lot in its desired place => deblock the algorithm
-        //find the first car that isn't in its right place
-        for (int i = 0; i < given.length; i++) {
-            if (given[i] != 0 && given[i] != desired[i]) {
-                return given[i];
-            }
+        // pick a random car from set of misplaced cars
+        if (misplacedCars.size() > 0) {
+            Integer nextCarToMove = (Integer)misplacedCars.toArray()[0];

```

Comment:
Good point, though the amortized complexity should be lower when we have a huge number of cars and just a few misplaced ones. If we change to a list / queue, removing elements from the list every time we put them in place will be O(N), as opposed to removing from a set. Any ideas?